fontforgeexe/multidialog.c

```diff
@@ -0,0 +1,642 @@
+/* -*- coding: utf-8 -*- */
+/* Copyright (C) 2000-2012 by George Williams, 2021 by Skef Iterum */
+/*
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+
+ * The name of the author may not be used to endorse or promote products
+ * derived from this software without specific prior written permission.
+
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+ * EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+ * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <fontforge-config.h>
+
+#include "basics.h"
+
+#include "ffglib.h"
+#include "fontforgeui.h"
+#include "multidialog.h"
+#include "../gdraw/ggadgetP.h"
+
+#include <assert.h>
+
+#define CID_BF_PAIR_START 0x10
+#define CID_STR_START 0x1000
+#define CID_LIST_START 0x4000
+
+enum multi_done_state { mds_not = 0, mds_ok = 1, mds_cancel = 2, mds_apply = 3 };
+
+static int Multi_Button(GGadget *g, GEvent *e) {
+    if (e->type == et_controlevent && e->u.control.subtype == et_buttonactivate) {
+        enum multi_done_state *mdsp = GDrawGetUserData(GGadgetGetWindow(g));
+        enum multi_done_state mds = (enum multi_done_state) GGadgetGetUserData(g);
+        *mdsp = mds;
+    }
+    return (true);
+}
+
+static int md_e_h(GWindow gw, GEvent *event) {
+    if (event->type == et_close) {
+        enum multi_done_state *mdsp = GDrawGetUserData(gw);
+        *mdsp = mds_cancel;
+    } else {
+        return false;
+    }
+    return true;
```

Comment:
I don't understand. Why can't we just drop the `else` and have a `return false` line?

Reply:
I'm just copying a common semantic from other handlers. 

Looking through the code it I can't find an easy example of an event handler return value being attended to, so maybe it's irrelevant? I have been assuming that `true` "consumes" the event and `false` means the processing might continue to higher or lower handlers. 

Can @jtanx offer any clarification?

Reply:
It's used e.g. here https://github.com/fontforge/fontforge/blob/e873b8f43e35bfc8d2053534c4e071a956115819/gdraw/gcontainer.c#L356

to determine whether or not an event was handled, so yes your interpretation is right, I think. The gcontainer stuff was always a bit complicated, so not entirely sure how much of it actually ends up being used/important though

Reply:
agree with dropping the redundant `else` though.

Reply:
Oh, I'm guessing this was my reminder to do something about the F1/Help link, but I didn't decide whether to link to the multidialog help page (potentially useful but counter-intuitive) or allow the caller to supply their own help somehow. 

I'll fix it one way or another.

Reply:
Resolved